src/realm/deppart/byfield_gpu_kernels.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+#include "realm/deppart/byfield.h"
+
+namespace Realm {
+
+// Intersect all instance rectangles with all parent rectangles in parallel.
+// Used for both count and emit depending on whether the output array is null.
+
+template <int N, typename T>
+__global__ void intersect_input_rects(
+  const SparsityMapEntry<N,T>* d_inst_entries,
+  const SparsityMapEntry<N,T>* d_parent_entries,
+  const size_t *d_inst_offsets,
+  const uint32_t *d_inst_prefix,
+  size_t numInstRects,
+  size_t numParentRects,
+  size_t numInsts,
+  uint32_t *d_inst_counters,
+  Rect<N,T>* d_rects
+) {
+  size_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= numInstRects * numParentRects) return;
+  size_t idx_x = idx % numParentRects;
+  size_t idx_y = idx / numParentRects;
+  assert(idx_x < numParentRects);
+  assert(idx_y < numInstRects);
+  const SparsityMapEntry<N, T> source_entry = d_parent_entries[idx_x];
+  const SparsityMapEntry<N, T> inst_entry = d_inst_entries[idx_y];
+  Rect<N,T> rect_output = inst_entry.bounds.intersection(source_entry.bounds);
+  if (rect_output.empty()) {
+    return;
+  }
+  size_t low = 0, high = numInsts;
+  while (low < high) {
+      size_t mid = (low + high) >> 1;
```

Comment:
look up and do the "no overflow" binary search. This can overflow if low and high are too big. It's unlikely to happen here because numInsts is small but just to use the "correct" algorithm.